user/languages/_includes/modern-c-cpp-toolchain.md

```diff
@@ -0,0 +1,130 @@
+### Modern C/C++ and Toolchain Versioning
+
+If you need a more modern toolchain or more granular control over your tools, you will have to handle that yourself. Fortunately, the community has already whitelisted the [repositories](https://github.com/travis-ci/apt-source-whitelist/blob/master/ubuntu.json) and [packages](https://github.com/travis-ci/apt-package-whitelist/blob/master/ubuntu-precise) necessary for most such upgrades.
+
+#### GCC on Linux
+
+Ubuntu 12.04 shipped with GCC 4.6.3 and Ubuntu 14.04 shipped with GCC 4.8.2. If you need a more recent version (if you are working with C++11, you *must* do this; support for C++11 was not finished until GCC 4.9), you will want to follow these directions:
+
+```yaml
+matrix:
+  include:
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - g++-4.9
+      env:
+         - MATRIX_EVAL="CC=gcc-4.9 && CXX=g++-4.9"
```

Comment:
The motivation was to, in some shape or form, standardize how the C++ compiler should be specified. Ideally I would've done

```
env:
  - CC=gcc-4.9
  - CXX=g++-4.9
```

but Travis overrides `CC` and `CXX` after the environment is configured per the build matrix, which forces the user to override them *again*, and the earliest point at which they can do so is in `before_install`. I'm reluctant to PR into Travis itself to support additional values for `compiler`, e.g. `gcc-4.9`, `gcc-5`, `gcc-6`, because (in my mind) it would cause a plethora of issues with ecosystem and images.

That all being said, I'm not exactly sure what you mean by "secondary concern" - could you clarify that, and what a preferable alternative would look like?